### ISSUE TITLE
Error with tensorflow 2:  Inputs to eager execution function cannot be Keras symbolic tensors

### ISSUE THREAD

**Issue by lutzgruber** [rstudio/keras#1008]
Hi everyone,

I am running into a problem migrating my code from tensorflow 1 to tensorflow 2. 

Below is sample code for a training a model with a custom loss function for a binomial distribution (with arbitrary and sample-specific N) that works with tensorflow 1 but throws an error with tensorflow 2 in `keras::fit`:

```
Error in py_call_impl(callable, dots$args, dots$keywords) : 
  _SymbolicException: Inputs to eager execution function cannot be Keras symbolic tensors, but found [<tf.Tensor 'logNchooseK:0' shape=(None, 1) dtype=float32>, <tf.Tensor 'N:0' shape=(None, 1) dtype=float32>]
```

I am using keras input layers in my custom loss function as my loss function depends on additional information other than the predictors and observed values. In this example, the outside information are the number of trials for each observation (N) and the appropriate binomial coefficient for the likelihood/loss function that cannot easily computed with the keras helper functions (logNchooseK).

Could you please advise on how to accomplish this with tensorflow 2?

```{r}
library(dplyr)

# create conda environments with tensorflow 1 and 2
reticulate::conda_create(envname = "tfpython", packages = c("tensorflow=1"))
reticulate::conda_create(envname = "tf2python", packages = c("tensorflow=2"))

k = list(
  # set this to tf2pthon to reproduce the error
  # set it to tfpython to reproduce the functioning example
  # TODO restart your R session before changing the conda environment
  tfenv = "tf2python",
  seed = 12345,
  validation_split = .2
)

set.seed(k$seed)
reticulate::use_condaenv(k$tfenv)

# define network -------------------------------------------------------------------------------------------

predictors_layer = keras::layer_input(
  shape = 2,
  name = "predictors"
)

N_layer = keras::layer_input(
  shape = 1,
  name = "N"
)

logNchooseK_layer = keras::layer_input(
  shape = 1,
  name = "logNchooseK"
)

output_layer = keras::layer_dense(
  predictors_layer,
  units = 1,
  activation = "linear",
  name = "output"
)

# define model --------------------------------------------------------------------------------------------

binomial_loss = function(y_true, y_pred) {
  predicted_prob = keras::k_exp(y_pred) / (1 + keras::k_exp(y_pred))
  
  neg_loglik = -(logNchooseK_layer + y_true * keras::k_log(predicted_prob) +
    (N_layer - y_true) * keras::k_log(1 - predicted_prob))
  
  return(neg_loglik)
}

model = keras::keras_model(
  inputs = c(predictors_layer, N_layer, logNchooseK_layer),
  outputs = output_layer
)

keras::compile(
  model,
  loss = binomial_loss,
  optimizer = keras::optimizer_adam()
)

# train model ----------------------------------------------------------------------------------------------

sample_data = tibble::tibble(
  var1 = rnorm(20),
  var2 = rnorm(20),
  N = sample(100, size = 20, replace = TRUE),
  true_prob = plogis(.5  * var1 + .75 * var2),
  outcome = rbinom(n = 20, size = N, prob = true_prob),
  logNchooseK = lchoose(N, outcome)
)

# this works with tensorflow 1 but crashes with tensorflow 2
keras::fit(
  model,
  x = list(
    predictors = sample_data %>%
      select(var1, var2) %>%
      mutate_all(as.double) %>%
      as.matrix(),
    N = sample_data %>%
      select(N) %>%
      mutate_all(as.double) %>%
      as.matrix(),
    logNchooseK = sample_data %>%
      select(logNchooseK) %>%
      mutate_all(as.double) %>%
      as.matrix()
  ),
  y = list(
    output = sample_data %>%
      select(outcome) %>%
      mutate_all(as.double) %>%
      as.matrix()
  ),
  validation_split = k$validation_split,
  view_metrics = TRUE,
  epochs = 1000
)

# evaluate model ------------------------------------------------------------------------------------------

predicted_probabilities = plogis(
  model$predict(
    x = list(
      predictors = sample_data %>%
        select(var1, var2) %>%
        mutate_all(as.double) %>%
        as.matrix(),
      N = sample_data %>%
        select(N) %>%
        mutate_all(as.double) %>%
        as.matrix(),
      logNchooseK = sample_data %>%
        select(logNchooseK) %>%
        mutate_all(as.double) %>%
        as.matrix()
    )
  )
)

plot(
  x = predicted_probabilities,
  y = sample_data$true_prob,
  col = ifelse(1:nrow(sample_data) < (1-k$validation_split) * nrow(sample_data), "red", "blue")
)
abline(a = 0, b = 1)
```

Your help is much appreciated. Thank you!

**Comment by lutzgruber** [comment]
It appears that setting `experimental_run_tf_function=FALSE` in the keras compile function as suggested in https://github.com/tensorflow/probability/issues/519 solves this problem for now. 

Could anyone please comment if this is a permanent solution or just a temporary fix?

**Comment by skeydan** [comment]
Hi,

I don't know about the future plans of the Google TensorFlow developers, but I could imagine this won't go away that easily. Perhaps an alternative would be rewriting your model.

It seems like the problem comes from the use of the "raw" input layers 2&3 in the loss function.
We can see this if we change it to

```
binomial_loss = function(y_true, y_pred) {
  predicted_prob = keras::k_exp(y_pred) / (1 + keras::k_exp(y_pred))
  
  #neg_loglik = -(logNchooseK_layer + y_true * keras::k_log(predicted_prob) +
  #                 (N_layer - y_true) * keras::k_log(1 - predicted_prob))
  
  #return(neg_loglik)
  predicted_prob
}
```

in which case, you don't get the error.

In the model, input layers 2&3 are not connected to anything and so, just reflect the raw input tensors. 
Perhaps you could try connecting them to lambda layers that don't do anything, like so

```
layer_lambda(f=function(x)x)
```

and connect each of them to a (dummy) output layer as well (so the output would be a list as well).
Then in the loss function, you would split up `y_true` and only use the parts needed in each place.

I think this might be worth giving a try.

**Comment by lutzgruber** [comment]
Thank you for your suggestion. I will wait for a resolution of my tensorflow issue and then see if workarounds as the one you proposed are necessary. In the meantime, I have a temporary solution with `experimental_run_tf_function=FALSE`.

I believe this is a core tensorflow issue as I could reproduce the error in Python, so I am closing this for now.
